Add sveltekit super rentals blog post (#1952)
mainmatter/mainmatter.com: src/posts/2022-11-18-sveltekit-super-rentals.md

---

> title: "A practical first look at the new Svelte framework"

Author: 👍

> authorHandle: mainmatter

Author: I know but I don't have twitter, should I just leave it out?

> authorHandle: mainmatter

Reviewer: This references your author file (see https://github.com/mainmatter/mainmatter.com/tree/master/src/authors) – I guess you can leave out Twitter there

SvelteJS - has been consistently popular since it emerged in 2019 (according to
the
[stateofjs polls](https://2021.stateofjs.com/en-US/libraries/front-end-frameworks))
and it looks like that popularity is only increasing.

And like all new JS frameworks, it’s also difficult to understand the benefits
and drawbacks of a new technology without getting hands-on with it, so that’s
exactly what this post intends to do.

The Svelte site has a nice
[Tutorial playground area](https://svelte.dev/tutorial/basics) that is useful to
understand the concepts behind the library, but because it is focused on the
component side of Svelte (as opposed to the framework, SvelteKit) there isn’t
any formal guide to help you get up and running with a real world application
using SvelteKit. Thankfully, EmberJS has a fairly thorough
[Tutorial](https://guides.emberjs.com/release/tutorial/part-1/) that is great
for introducing the principles of the framework. So we will be cross-pollinating
and using SvelteKit to build the Ember-Super-Rentals demo app. And while we’re
at it, we can also draw some comparisons between the frameworks.

> 

Author: Am I okay to use the screenshot from the ember site here? Or should I save it and add it myself?

> 

Reviewer: I guess you can use the one from the Ember site

> 

Author: Ignore this, I took a screenshot anyway and will use that

I have broken this demo app down into the same chunks as is done in the Ember
Super Rentals tutorial, so feel free to follow along with both tutorials at the
same time.

You can also find snapshots of this tutorial as separate tags in the

> [github repo](https://github.com/beerinho/sveltekit-super-rentals/tags).

Author: Done

## Let’s get started

> Before we get started, I just want to reiterate the notices on the SvelteKit
> documentation site. **This technology is not yet suggested for use in
> production applications.** SvelteKit is still pre version 1.0 and so is
> constantly changing large parts of the framework; you can see the progress and
> upcoming changes on the

> > [GitHub Discussions tab](https://github.com/sveltejs/kit/discussions?discussions_q=sort%3Atop).

Author: You're right, this shouldn't be a quote, I was using the quote block in Notion and couldn't see anything that helps to emphasise a section. Do you have any suggestions?

> > [GitHub Discussions tab](https://github.com/sveltejs/kit/discussions?discussions_q=sort%3Atop).

Reviewer: See my comment above – I think the only things we have that work here are italic or bold paragraphs. Using quotes is confusing – I was asking myself who you're quoting here until I realized it's actually the author speaking here still :)

> > [GitHub Discussions tab](https://github.com/sveltejs/kit/discussions?discussions_q=sort%3Atop).

Author: I've replaced this with `hr`s above and below to make it stand out more

 create mode 100644 src/app.html
 create mode 100644 src/routes/+page.svelte
 create mode 100644 static/favicon.png
 create mode 100644 svelte.config.js
 create mode 100644 tests/test.js
 create mode 100644 vite.config.js
svelte-super-rentals % npm run dev -- --open

> svelte-super-rentals@0.0.1 dev
> vite dev "--open"

  VITE v3.0.9  ready in 2475 ms

  ➜  Local:   http://localhost:5173/
  ➜  Network: use --host to expose
```

## PART 1

### Orientation:
[Repo](https://github.com/beerinho/sveltekit-super-rentals/tree/1.1-orientation)

To keep this simple, we’re going to use the styles and assets from the
Ember-Super-Rentals app and just copy them over into our project. I won’t paste
them in here as the style sheet is pretty big and not worth the page space. If
you want to see if it, feel free to check out the repo on GitHub.

Tomster image -
[https://guides.emberjs.com/downloads/teaching-tomster.png](https://guides.emberjs.com/downloads/teaching-tomster.png)

Styles -
[https://github.com/ember-learn/super-rentals/blob/super-rentals-tutorial-output/app/styles/app.css](https://github.com/ember-learn/super-rentals/blob/super-rentals-tutorial-output/app/styles/app.css)

> And when we serve the app we are greeted by the friendly face of Tomster but it

Author: Nice idea, added ✅

> This confused me at first as Playwright will stop at each `await` so you will
> need to manually step through the test until you find the part that you want
> to stop it. It is quite intuitive once you realise it might not have paused
> where you expected to begin with.

To open the inspector when running the tests, be sure to specify the `PWDEBUG=1`
flag when running the tests:

```js
PWDEBUG=1 npm run test
```

### Component basics:
[Repo](https://github.com/beerinho/sveltekit-super-rentals/tree/1.4-component-basics)

Similar to the routes, there is no command to generate components for us, so we
will create the `src/components` folder along with the
`src/components/jumbo.svelte` and `src/components/nav-bar.svelte` component
files

```js
// components/jumbo.svelte

<div class="jumbo">
  <div class="right tomster" />
  <slot />
</div>
```

The `<slot/>` here is similar to the `{{yield}}` in Ember, it allows us to use
this component as a wrapper for more HTML to be passed in from the parent.

```js
// components/nav-bar.svelte

<nav class="menu">
  <a href="/" class="menu-index">
    <h1>SuperRentals</h1>
  </a>
  <div class="links">
    <a href="/about" class="menu-about">

>       {" "}

Author: These must have just come from copying from Notion, I've removed them now

opted for the single file approach.

In terms of component usage, both Ember and Svelte have a similar approach but
slightly differing syntax, both have opted for more native looking HTML - Ember
using HBS, while Svelte goes for pure HTML that can take pure JS expressions.
Coming from Ember I found this aspect of Svelte to be quite easy to get to grips
with, though the naming of component attributes can get a little confusing as
there is no `@` syntax that Ember uses, so instead of `@image` you would have
`image`. I’ll discuss this more in the next section “More about components”.

Another big difference between the two is that components are always available
in Ember and you can simply invoke them in the template, whereas in SvelteKit
you need to import the component in the script tag before you can use it in the
template. I like the simplicity of Ember’s approach, but it does mean you can
end up with quite long component names if your app has a lot of component
nesting - i.e. `<Ui::Layout::Foo::Bar::TwoColumn::AwesomeComponent/>` whereas
you likely don’t have this problem in Svelte because you can simply change the

> name of the component when you import it - i.e.

Reviewer: I'm not entirely sure and there might be multiple relevant RFCs but I guess that's one of them, yes.

```js
<script>
	import AwesomeComponent from '@components/ui/layout/foo/bar/two-column/awesome-component.svelte'
</script>

<AwesomeComponent/>
```

But it can be argued that Svelte’s approach could lead to more confusion as the
names can change depending on how they are imported, meaning it’s not

> immediately obvious which component is which.

Author: I don't think named exports could work here, but the components can certainly be imported as whatever you like. I've added a note about that and the lint rule 👍

We also created the NavBar, which we want to be visible on all pages, instead of
adding it individually to each page component, we will create a `+layout.svelte`
component instead. The layout component sits in the nesting structure just like
the `pages` and will apply a layout to all children routes unless otherwise
specified.

```js
// routes/+layout.svelte

<script>
	import NavBar from '@components/nav-bar.svelte';
</script>

<NavBar />
<slot />
```

As mentioned, this will apply the `<NavBar/>` to all pages, the `<slot/>` is
required and is similar to the `{{outlet}}` in Ember, in which it will allow a

> portal to render the route content through.

Author: I'm not sure of the best way to phrase this, but it is how the `{{outlet}}` and `{{yield}}` blocks work. I've updated it to hopefully be a bit clearer

to using the test generator was that it wasn’t so simple to test that something
was visible, to get around this it required clicking on a non-interactive
element and then adjusting the test to test the content rather than just
clicking on the element.

### More about components
[Repo](https://github.com/beerinho/sveltekit-super-rentals/tree/1.5-more-about-components)

Now let’s create some components that will receive parameters to be handled by
the component.

```js
// components/rental/image.svelte

<script>
	export let src;
	export let alt;
</script>

<div class="image">
	<img {src} {alt} />
</div>
```

A big difference between Ember and SvelteKit here is that Svelte requires you to
express which parameters are expected by the component using the `export let`
syntax. Coming from a JS background, it’s a little confusing to use `export` in
this way, but in general I prefer having to define the parameters over glimmers
way of have the `this.args.` object that can be of any shape, though this can
get confusing in Svelte when you have bigger components with more internal as
well as external attributes, with Svelte it isn’t always clear if the attribute

> is external or internal, whereas the `@/this` syntax used in Ember gives a

Author: This is supposed to mean the `@` or `this`, so I have updated it to hopefully be a little clearer

> is external or internal, whereas the `@/this` syntax used in Ember gives a

Reviewer: ah, right – `@this` doesn't exist anyway :)

{/if}
```

This syntax will look very similar to that of Ember, with a few minor changes,
instead of relying on Glimmer’s `{{on}}` modifier, we user Svelte’s `on:`
modifier. Then instead of using an `@action` to link the template to the JS as
we would in Ember, we simply call a regular JS function.

But as you can see, the conditional syntax is very similar, where Ember has

```js
{{#if this.thing}}
	show thing
{{else}}
	show other thing
{{/if}}
```

Svelte has

```js
{#if thing}
	show thing
{:else}
	show other thing
{/if}
```

### Reusable components
[Repo](https://github.com/beerinho/sveltekit-super-rentals/tree/1.7-reusable-components)

Now let’s add an external API, for this we will need a Mapbox account and to
create an access token. You can do so [here](https://www.mapbox.com/signup/).

Because we don’t have the full `environment` config that is present in Ember
applications, we will need to store our Mapbox access token elsewhere. We could
do this in a number of places (e.g. in a store, or in the component) but for
this example, we will use the suggested route, we will create a `.env` file in
the root of our project and add `VITE_MAPBOX_ACCESS_TOKEN = 'your_access_token’`

> (this is ignored by git so isn’t shown in the repo), it is also important to

Reviewer: Easiest fix might be to just include it in the repo? The token ends up being published as part of the site anyway so doesn't need to be kept a secret?

> (this is ignored by git so isn’t shown in the repo), it is also important to

Reviewer: > required and is similar to the `{{outlet}}` in Ember, which allows the parent to render content to a specific area of the child component.

I think it's the other way round:

> required and is similar to the `{{outlet}}` in Ember, which allows child routes to render content into a specific area of the parent route.

> (this is ignored by git so isn’t shown in the repo), it is also important to

Author: 🤦

There are a couple of things to unwrap in this component, the first is that we
have two `script` tags, the first script tag has `context="module"` which means
it is only instantiated once in the application and the variables will not
change. This is useful because the `token` and `MAPBOX_API` attributes will be
the same for every instantiation of the component within our app.

The second `script` is not a module which means the properties within it can
change between instances, this is important as we will be using this component
in multiple places to show the different locations of our rentals.

The next big difference is the `$` - this defines a computed property which will
be updated and cause a re-render in the component when it changes. This is very
similar to a `get` in Ember or a `@tracked` property in Glimmer.

So in Ember this line would look like:

```js
get src(){
	return `${MAPBOX_API}/${this.args.lng},${this.args.lat},${this.args.zoom}/${this.args.width}x${this.args.height}@2x?access_token=${token}`
}
```

or

```js
@tracked
	src = `${MAPBOX_API}/${this.args.lng},${this.args.lat},${this.args.zoom}/${this.args.width}x${this.args.height}@2x?access_token=${token}`

> ```

Author: I mention afterwards that this isn't actually a very good pattern for this as the attributes don't change after the component is rendered, but I just wanted to replicate what is in the Ember tutorial. I'm fine with removing it if you think it's confusing

> ```

Reviewer: ok, if this is in the Ember guides I guess it's fine

```js
// components/rental/index.svelte

<script>
	import RentalImage from './image.svelte';
	import Map from '@components/map.svelte';

+	export let rental
</script>

<article class="rental">
-	<RentalImage
-		src="https://upload.wikimedia.org/wikipedia/commons/c/cb/Crane_estate_(5).jpg"
-		alt="A picture of Grand Old Mansion"
-	/>
+	<RentalImage src={rental.image} alt={rental.description} />
	<div class="details">
-		<h3>Grand Old Mansion</h3>
+		<h3>{rental.title}</h3>
		<div class="detail owner">
-			<span>Owner:</span> Veruca Salt
+			<span>Owner:</span> {rental.owner}
		</div>
		<div class="detail type">
-			<span>Type:</span> Standalone
+			<span>Type:</span> {rental.type}
		</div>
		<div class="detail location">
-			<span>Location:</span> San Francisco
+			<span>Location:</span> {rental.city}
		</div>
		<div class="detail bedrooms">
-			<span>Number of bedrooms:</span> 15
+			<span>Number of bedrooms:</span> {rental.bedrooms}
		</div>
	</div>
	<Map
-		lat="37.7749"
-		lng="-122.4194"
+		lat={rental.location.lat}
+		lng={rental.location.lng}
		zoom="9"
		width="150"
		height="150"
-		alt="A map of Grand Old Mansion"
+		alt="A map of {rental.title}"
	/>
</article>
```

> And now we have a lovely looking index page that shows us the data for each of

Author: ✅

      <span>Type:</span>
      {rental.type} – {rental.category}
    </div>
    <div class="detail location">
      <span>Location:</span>
      {rental.city}
    </div>
    <div class="detail bedrooms">
      <span>Number of bedrooms:</span>
      {rental.bedrooms}
    </div>
    <div class="detail description">
      <p>{rental.description}</p>
    </div>
  </div>

  <Map
    lat={rental.location.lat}
    lng={rental.location.lng}
    zoom="12"
    width="894"
    height="600"
    alt="A map of {rental.title}"
    styleClass="large"
  />
</article>
```

As before, we’ll use a `script` tag with `context=”module”` to import things
that won’t change between instances of the page, and a `script` without the
context for the attributes that can change. You’ll notice here that we are
expanding the `data` object and assigning `data.rental` to it’s own property to
make it easier to reference in the template. We will implement the “Share on
Twitter” button later so this is just a placeholder for now.

We also need to be able to add a class to the `Map` component so that it can be
styled from the parent, so to do this we will add a new attribute to the
component.

> ```js

Reviewer: ```suggestion
```diff
```

```js
// tests/test.js

...
test('viewing the details of a rental property', async ({ page }) => {
	// Go to the index route
	await page.goto('/');
	// Assert there there 3 properties displayed
	await expect(page.locator('.rental')).toHaveCount(3);

	// Click on a rental
	await page.locator('text=Grand Old Mansion').click();
	// Assert it has taken you to the rental page
	await expect(page).toHaveURL('/rentals/grand-old-mansion');
});

test('visiting /rentals/grand-old-mansion', async ({ page }) => {
	// Go to the rental page
	await page.goto('/rentals/grand-old-mansion');
	// Assert we are on the current URL
	await expect(page).toHaveURL('/rentals/grand-old-mansion');

	// Assert the nav exists
	await expect(page.locator('nav')).toBeVisible();
	// Assert the main title is still correct
	expect(await page.textContent('h1')).toBe('SuperRentals');
	// Assert the page title has the correct text
	expect(await page.textContent('h2')).toBe('Grand Old Mansion');
	// Assert the rental detail area is visible
	await expect(page.locator('.rental.detailed')).toBeVisible();
});
```

### Service injection
[Repo](https://github.com/beerinho/sveltekit-super-rentals/tree/2.2-service-injection)

> Let’s start off by replacing that placeholder “Share to Twitter” button we

Reviewer: more like

```suggestion
Let’s continue by replacing that placeholder “Share to Twitter” button we
```

?

to Twitter. First, we’ll create a new component to house all of this Twitter
logic

```js
// components/share-button.svelte

<script>
	import { onMount } from 'svelte';

	const TWEET_INTENT = 'https://twitter.com/intent/tweet';

	export let text
	export let hashtags
	export let via

	let shareURL;

	onMount(() => {
		let url = new URL(TWEET_INTENT);

		url.searchParams.set('url', window.location.href);

		if (text) {
			url.searchParams.set('text', text);
		}

		if (hashtags) {
			url.searchParams.set('hashtags', hashtags);
		}

		if (via) {
			url.searchParams.set('via', via);
		}

		shareURL = url.href;
	});
</script>

<a href={shareURL} target="_blank" rel="external nofollow noopener noreferrer" class="share button">
	<slot />
</a>
```

> The `onMount` function here is very similar to Glimmer’s `constructor`, this
> is where you store the logic that needs to run before the component is
> rendered, in this scenario we will be populating the URL that we will need to
> create a tweet to share the rental we are looking at.

We will replace the placeholder with this new component on the `rentals/[slug]`
page

> ```js

Reviewer: ```suggestion
```diff
```

});

test('the index page updates the results according to the search query', async ({ page }) => {
	// Go to the index page
	await page.goto('/');
	// Assert we are on the current URL
	await expect(page).toHaveURL('/');

	// assert all rentals are loaded correctly
	expect(page.locator('.rentals')).toBeVisible();
	expect(page.locator('.rentals input')).toBeVisible();
	expect(page.locator('.rentals .results')).toBeVisible();
	expect(page.locator('.rentals .results li')).toHaveCount(3);

	// Fill input
	await page.locator('input').fill('Mansion');

	// assert that there is only one result and that it's title includes the word "mansion"
	expect(page.locator('.rentals .results li')).toHaveCount(1);
	expect(await page.textContent('.rentals .results li h3')).toBe('Grand Old Mansion');

	// Fill input with new search term
	await page.locator('input').fill('DownTown');

	// assert that there is only one result and that it's title includes the word "downtown"
	expect(page.locator('.rentals .results li')).toHaveCount(1);
	expect(await page.textContent('.rentals .results li h3')).toBe('Downtown Charm');
});
```

> And there we have it, SvelteKit Super Rentals!

Reviewer: Would be cool to link to a deployment of the actual app on Netlify or sth.

Author: I was thinking of creating some follow-up posts integrating firebase hosting, authentication, CRUD, and cloud functions so would add it then, I could do this one separately though?

Reviewer: I meant just link to a deployment of the app in the state it would be in at this point in the tutorial. I think you could simply set up Netlify deployment for [this repo](https://github.com/mainmatter/sveltekit-super-rentals) and that should just work?

Author: Done, I also added a link to it at the top of the post.

### Final Thoughts

I really enjoyed getting stuck in with SvelteKit and I can certainly see the
appeal of it, it’s a very lightweight framework and in a lot of ways it stays

> very close to native javascript, which makes me think I think it would be a

Reviewer: ```suggestion
very close to native javascript, which makes me think it would be a
```

image: ""
imageAlt: ""
---

{% raw %} As with all new JS frameworks, there is a big buzz surrounding
SvelteKit, and for good reason, it’s still pre-1.0 but the underlying library -
SvelteJS - has been consistently popular since it emerged in 2019 (according to
the
[stateofjs polls](https://2021.stateofjs.com/en-US/libraries/front-end-frameworks))
and it looks like that popularity is only increasing.

And like all new JS frameworks, it’s also difficult to understand the benefits
and drawbacks of a new technology without getting hands-on with it, so that’s
exactly what this post intends to do.

The Svelte site has a nice
[Tutorial playground area](https://svelte.dev/tutorial/basics) that is useful to
understand the concepts behind the library, but because it is focused on the
component side of Svelte (as opposed to the framework, SvelteKit) there isn’t
any formal guide to help you get up and running with a real world application
using SvelteKit. Thankfully, EmberJS has a fairly thorough
[Tutorial](https://guides.emberjs.com/release/tutorial/part-1/) that is great

> for introducing the principles of the framework. So we will be cross-pollinating

Author: I think these sentences need to be combined into one as it's still talking about the same thing

        at process.<anonymous> (/Users/dan/Documents/Repos/svelte-super-rentals/node_modules/.pnpm/@playwright+test@1.21.1/node_modules/@playwright/test/lib/worker.js:94:5)

    Pending operations:
      - page.textContent at tests/test.ts:6:21

  Slow test file: tests/test.ts (30s)
  Consider splitting slow test files to speed up parallel execution

  1 failed
    tests/test.ts:2:1 › about page has expected h1 =================================================
```

As we would expect, this test is failing (after a while) because we are
expecting to find an `h1` on the page with the text “About this app”, which we
no longer have since we replaced the default content of the file with the HTML
for the Super Rentals app.

Now let’s update the test so that is passes.

```js
// tests/test.js

import { expect, test } from "@playwright/test";

test("about page has expected h2", async ({ page }) => {
  await page.goto("/about");
  expect(await page.textContent("h2")).toBe("About Super Rentals");
});
```

And as expected, the test will now pass.

> ```js

Reviewer: ```suggestion
```bash
```

```js
await page.pause();
```

To resume the tests, open up the devTools console in the browser and run this
command:

```js
playwright.resume();
```

Otherwise you can use the
[Playwright inspector](https://playwright.dev/docs/inspector) to open a browser
along with a more in-depth console that will give you the ability to step
through your tests and see more detail about the current state of the DOM and
test.

---

This confused me at first as Playwright will stop at each `await` so you will
need to manually step through the test until you find the part that you want to
stop it. It is quite intuitive once you realise it might not have paused where
you expected to begin with.

---

To open the inspector when running the tests, be sure to specify the `PWDEBUG=1`
flag when running the tests:

```js
PWDEBUG=1 npm run test
```

### Component basics:

[Repo](https://github.com/mainmatter/sveltekit-super-rentals/tree/1.4-component-basics)

Similar to the routes, there is no command to generate components for us, so we
will create the `src/components` folder along with the
`src/components/jumbo.svelte` and `src/components/nav-bar.svelte` component
files

> ```js

Reviewer: ```suggestion
```html
```

// components/jumbo.svelte

<div class="jumbo">
  <div class="right tomster" />
  <slot />
</div>
```

The `<slot/>` here is similar to the `{{yield}}` in Ember, it allows us to use
this component as a wrapper for more HTML to be passed in from the parent.

> ```js

Reviewer: ```suggestion
```html
```

// components/nav-bar.svelte

<nav class="menu">
  <a href="/" class="menu-index">
    <h1>SuperRentals</h1>
  </a>
  <div class="links">
    <a href="/about" class="menu-about">
      About
    </a>
    <a href="/getting-in-touch" class="menu-contact">
      Contact
    </a>
  </div>
</nav>
```

I then added the alias to the svelte config to make importing files easier -
this means instead of having a mess of `../../../..` in the module imports, we
can use `@components` when importing a component into another file

```js
// svelte.config.js

import adapter from "@sveltejs/adapter-auto";
import path from "path";

/** @type {import('@sveltejs/kit').Config} */
const config = {
  kit: {
    adapter: adapter(),
    // to make it easier to handle module imports, we alias them here
    alias: {
      "@components": path.resolve("./src/components"),
    },
  },
};

export default config;
```

Then we will update our usage of these components in our routes

> ```js

Reviewer: ```suggestion
```html
```

// routes/+page.svelte

<script>
  import Jumbo from '@components/jumbo.svelte';
</script>

<Jumbo>
  <h2>Welcome to Super Rentals!</h2>
  <p>
    We hope you find exactly what you're looking for in a place to stay.
  </p>
  <a href="/about" class="button">About Us</a>
</Jumbo>
```

> ```js

Reviewer: ```suggestion
```html
```

//routes/about/+page.svelte

<script>
  import Jumbo from '@components/jumbo.svelte'
</script>

<Jumbo>
  <h2>About Super Rentals</h2>
  <p>
    The Super Rentals website is a delightful project created to explore Ember. By building a
    property rental site, we can simultaneously imagine traveling AND building Ember applications.
  </p>
  <a href="/getting-in-touch" class="button">Contact Us</a>
</Jumbo>
```

> ```js

Reviewer: ```suggestion
```html
```

// routes/getting-in-touch/+page.svelte

<script>
  import Jumbo from '@components/jumbo.svelte'
</script>

<Jumbo>
  <h2>Contact Us</h2>
  <p>
    Super Rentals Representatives would love to help you<br />
    <a href="mailto:superrentalsrep@sveltedemo.com">superrentalsrep@sveltedemo.com</a>
  </address>
  <a href="/about" class="button">About</a>
</Jumbo>
```

As you can see, another difference between SvelteKit and Ember is that Ember has
chosen to separate the template from the component logic, whereas Svelte has
opted for the single file approach (although this is currently in an
[RFC](https://github.com/emberjs/rfcs/blob/master/text/0779-first-class-component-templates.md#sfcs)

> to be added to Ember).

Reviewer: The RFC has been accepted and merged already and I think the feature even has been implemented and this should actually be released soon.

In terms of component usage, both Ember and Svelte have a similar approach but
slightly differing syntax, both have opted for more native looking HTML - Ember

> using HBS, while Svelte goes for pure HTML that can take pure JS expressions.

Reviewer: > goes for pure HTML

I'm not sure that's true? It's a template language as well but one that allows JS expressions (essentially like JSX)?

Author: I've updated this but still not sure of the best way to have a succinct explanation of it. Essentially you can write native HTML and add JS to it. Not quite like JSX though as that is more HTML through JS, whereas with Svelte you can have HTML without the JS

Coming from Ember I found this aspect of Svelte to be quite easy to get to grips
with, though the naming of component attributes can get a little confusing as
there is no `@` syntax that Ember uses, so instead of `@image` you would have
`image`. I’ll discuss this more in the next section “More about components”.

Another big difference between the two is that components are always available
in Ember and you can simply invoke them in the template, whereas in SvelteKit
you need to import the component in the script tag before you can use it in the
template. I like the simplicity of Ember’s approach, but it does mean you can
end up with quite long component names if your app has a lot of component
nesting - i.e. `<Ui::Layout::Foo::Bar::TwoColumn::AwesomeComponent/>` - whereas
you likely don’t have this problem in Svelte because you can simply change the
name of the component when you import it - i.e.

> ```js

Reviewer: ```suggestion
```html
```

<script>
  import AwesomeComponent from '@components/ui/layout/foo/bar/two-column/awesome-component.svelte'
</script>

<AwesomeComponent/>
```

---

> _Note: There is currently an open

Reviewer: See above, the RFC isn't open – it has been accepted and merged and even implemented already.

[RFC](https://github.com/emberjs/rfcs/blob/master/text/0779-first-class-component-templates.md)
which proposes to add template imports to Ember._

---

But it can be argued that Svelte’s approach could lead to more confusion as the
names can change depending on how they are imported, meaning it’s not

> immediately obvious which component is which; although this isn't the strongest

Author: I'm not sure about this one, it doesn't feel like the second clause here needs any extra emphasis as it's more of a side note

argument as this issue could easily be mitigated by using good naming
conventions or implementing a lint rule for this.

We also created the `<NavBar>`, which we want to be visible on all pages,
instead of adding it individually to each page component, we will create a
`+layout.svelte` component instead. The layout component sits in the nesting
structure just like the `pages` and will apply a layout to all children routes
unless otherwise specified.

> ```js

Reviewer: ```suggestion
```html
```

```

As mentioned, this will apply the `<NavBar/>` to all pages. The `<slot/>` is
required and is similar to the `{{outlet}}` in Ember, which allows child routes
to render content into a specific area of the parent route.

If we wanted a different layout for just the `about` page, we could create
another layout in `routes/about/+layout.svelte` and that would not impact the
other routes, but would apply a separate layout to all routes nested with
`/about`.

SvelteKit doesn’t come with a library for testing standalone components but we
did opt to include Playwright when setting up the project, so I initially wanted
to implement Playwright’s experimental Component Testing library but as of the
time of writing this, there are issues with the library that mean I wasn’t able
to use it in this project. (there are other options for unit testing including
[Vitest](https://vitest.dev/guide/) but for the sake of brevity I’ll just stick
to e2e testing using Playwright for this example)

Instead of writing out the code, let’s try and use
[Playwright’s test generator](https://playwright.dev/docs/codegen) with the
command:

_Note: make sure your app is already running on the port specified_

> ```js

Reviewer: ```suggestion
```bash
```

```js
// Go to the index route
await page.goto("/");
// Click nav >> text=About
await page.locator("nav >> text=About").click();
await expect(page).toHaveURL("/about");
// Click nav >> text=Contact
await page.locator("nav >> text=Contact").click();
await expect(page).toHaveURL("/getting-in-touch");
// Click h1:has-text("SuperRentals")
await page.locator('h1:has-text("SuperRentals")').click();
await expect(page).toHaveURL("/");
```

While using the Playwright test generator, I loved the simplicity of being able
to use the UI to test the things that I wanted assure was actually working in
the UI, so from this aspect I found it really useful. The main downside I found
to using the test generator was that it wasn’t so simple to test that something
was visible, to get around this it required clicking on a non-interactive
element and then adjusting the test to test the content rather than just
clicking on the element.

### More about components

[Repo](https://github.com/mainmatter/sveltekit-super-rentals/tree/1.5-more-about-components)

Now let’s create some components that will receive parameters to be handled by
the component.

> ```js

Reviewer: ```suggestion
```html
```

// components/rental/image.svelte

<script>
  export let src;
  export let alt;
</script>

<div class="image">
  <img {src} {alt} />
</div>
```

A big difference between Ember and SvelteKit here is that Svelte requires you to
express which parameters are expected by the component using the `export let`
syntax. Coming from a JS background, it’s a little confusing to use `export` in
this way, but in general I prefer having to define the parameters over glimmers

> way of have the `this.args.` object that can be of any shape, though this can

Reviewer: ```suggestion
way of have the `this.args` object that can be of any shape, though this can
```

get confusing in Svelte when you have bigger components with more internal as
well as external attributes, with Svelte it isn’t always clear if the attribute
is external or internal, whereas the `@`/`this` syntax used in Ember gives a
clearer distinction here.

`<img {src} {alt} />` is making use of Svelte’s shorthand mechanism, if the

> attribute you are setting has the same name as the parameter, you can omit the

Reviewer: ```suggestion
attribute you are setting has the same name as the variable, you can omit the
```

element attribute and Svelte will figure it out for you. This could also be
expressed as `<img src={src} alt={alt} />`.

Using the new `<RentalImage>` component, let’s create the `<Rental>` component,
which will just be hardcoded for now to give is some information to see on
screen.

> ```js

Reviewer: ```suggestion
```html
```

// components/rental/index.svelte

<script>
  import RentalImage from './image.svelte';
</script>

<article class="rental">
  <RentalImage
    src="https://upload.wikimedia.org/wikipedia/commons/c/cb/Crane_estate_(5).jpg"
    alt="A picture of Grand Old Mansion"
  />
  <div class="details">
    <h3>Grand Old Mansion</h3>
    <div class="detail owner">
      <span>Owner:</span> Veruca Salt
    </div>
    <div class="detail type">
      <span>Type:</span> Standalone
    </div>
    <div class="detail location">
      <span>Location:</span> San Francisco
    </div>
    <div class="detail bedrooms">
      <span>Number of bedrooms:</span> 15
    </div>
  </div>
</article>
```

And then add a few of these to the `index` page of our app.

> ```diff-js

Reviewer: ```suggestion
```diff-html
```

Author: This has a similar issue to the js, the diff styling can be added, but the HTML styling is missing

// routes/+page.svelte

<script>
  import Jumbo from '@components/jumbo.svelte';
+ import Rental from '@components/rental/index.svelte';
</script>

<Jumbo>
  <h2>Welcome to Super Rentals!</h2>
  <p>We hope you find exactly what you're looking for in a place to stay.</p>
  <a href="/about" class="button">About Us</a>
</Jumbo>

+<div class="rentals">
+  <ul class="results">
+    <li><Rental /></li>
+    <li><Rental /></li>
+    <li><Rental /></li>
+  </ul>
+</div>
```

### Interactive components

[Repo](https://github.com/mainmatter/sveltekit-super-rentals/tree/1.6-interactive-components)

Now let’s add the ability to toggle the size of the image on click.

> ```js

Reviewer: ```suggestion
```html
```

  show thing
{{else}}
  show other thing
{{/if}}
```

Svelte has

```js
{#if thing}
  show thing
{:else}
  show other thing
{/if}
```

### Reusable components

[Repo](https://github.com/mainmatter/sveltekit-super-rentals/tree/1.7-reusable-components)

Now let’s add an external API, for this we will need a Mapbox account and to
create an access token. You can do so [here](https://www.mapbox.com/signup/).

Because we don’t have the full `environment` config that is present in Ember
applications, we will need to store our Mapbox access token elsewhere. We could
do this in a number of places (e.g. in a store, or in the component) but for
this example, we will use the suggested route, we will create a `.env` file in
the root of our project and add `PUBLIC_MAPBOX_TOKEN = 'your_access_token’`, it
is also important to start the key with `PUBLIC_` so that SvelteKit is aware
that this token is public and will bundle it with all other static public keys
that can be accessed through `'$env/static/public'`(for more information, read
[here](https://kit.svelte.dev/docs/modules#$env-static-public)).

Now that we have the access token stored in our app, we can use it in our new
`<Map>` component

> ```js

Reviewer: ```suggestion
```html
```

// components/map.svelte

<script context="module">
  import { PUBLIC_MAPBOX_TOKEN }  from '$env/static/public';
  const MAPBOX_API = 'https://api.mapbox.com/styles/v1/mapbox/streets-v11/static';
</script>

<script>
  export let lat
  export let lng
  export let height
  export let width
  export let zoom
  export let alt = `Map image at coordinates ${lat},${lng}`;

  $: src = `${MAPBOX_API}/${lng},${lat},${zoom}/${width}x${height}@2x?access_token=${PUBLIC_MAPBOX_TOKEN}`;
</script>

<div class="map">
  <img {alt} {src} {width} {height} />
</div>
```

---

As you can see, the `img` tag uses the shorthand that was explained earlier in
the “More about components” section.

---

There are a couple of things to unwrap in this component, the first is that we
have two `script` tags, the first script tag has `context="module"` which means
it is only instantiated once in the application and the variables will not
change. This is useful because the `token` and `MAPBOX_API` attributes will be
the same for every instantiation of the component within our app.

> The second `script` is not a module which means the properties within it can

Author: I want it to be clear here that there is an order, so I've added the punctuation but kept the "first" and "second"

change between instances, this is important as we will be using this component
in multiple places to show the different locations of our rentals.

The next big difference is the `$` - this defines a computed property which will
be updated and cause a re-render in the component when it changes. This is very
similar to a `get` in Ember or a `@tracked` property in Glimmer.

So in Ember this line would look like:

```js
get src(){
  return `${MAPBOX_API}/${this.args.lng},${this.args.lat},${this.args.zoom}/${this.args.width}x${this.args.height}@2x?access_token=${PUBLIC_MAPBOX_TOKEN}`
}
```

or

```js
@tracked
  src = `${MAPBOX_API}/${this.args.lng},${this.args.lat},${this.args.zoom}/${this.args.width}x${this.args.height}@2x?access_token=${PUBLIC_MAPBOX_TOKEN}`
```

_(I noticed that none of this was actually required as the attributes don’t ever
change after the component is rendered, so this could be a variable on the
component but to keep this tutorial in line with the Ember Super Rentals
tutorial, we’ll keep it this way.)_

We can then add this new Map component to the bottom of the `article` on our
`<Rental>` component:

> ```js

Reviewer: ```suggestion
```html
```

Finally, let’s update our index page test to expect to see three of these
components

```js
// tests/test.js

...
test('visiting /', async ({ page }) => {
  ...
  // assert the title of the page is correct
  expect(await page.textContent('h2')).toBe('Welcome to Super Rentals!');
  // assert the map images are visible
  await expect(page.locator('.map img')).toHaveCount(3);
  ...
});
```

### Working with Data

[Repo](https://github.com/mainmatter/sveltekit-super-rentals/tree/1.8-working-with-data)

Now we have some nice looking components but they’re not very interesting, the
component is hardcoded with the rental information which means we just have
three of the same thing visible. So lets start populating our app with some
dynamic data - usually this would be populated from a server, but we will be
sticking with the Ember-super-rentals tutorial and store it inside our app as
JSON files, you can download the required files
[here](https://guides.emberjs.com/downloads/data.zip).

As this data isn’t going to change, and we want it to be available in our app
build, I’m going to store it inside the `static` directory in a new folder

> called `api` alongside or `images` folder. (You can call this folder whatever

Reviewer: ```suggestion
called `api` alongside our `images` folder. (You can call this folder whatever
```

or I guess

```suggestion
called `api` alongside the `images` folder. (You can call this folder whatever
```

you like, “api” isn’t a special name here), then inside the folder we will have
our rentals data:

> ```xml

Reviewer: ```suggestion
```bash
```

static/
  api/
    rentals.json
    rentals/
      downtown-charm.json
      grand-old-mansion.json
      urban-living.json
```

I’ve used the slugs of each rental as the name to make it easier to reference
and retrieve the file when we “load” the data.

As with the `@components` alias, I’ll add the same for the api files to make
them easier to reference.

> ```xml

Reviewer: ```suggestion
```js
```

  alias: {
    '@api': path.resolve('./static/api'),
    '@components': path.resolve('./src/components')
  }
  ...
```

Now we can introduce one of the strengths of SvelteKit, the `+page.server.js`
file. This can be added to any page in your application that needs to trigger
something on the server - usually loading/saving data. And this is only run on
the server, which means you can expose sensitive credentials here without the
client ever knowing about them. We don’t have any sensitive information in our
app but it also gives us the benefit of having all of the data loaded and in the
page source which means it is search engine bots are able to crawl it easier,
increasing it’s SEO score.

```js
// routes/+page.server.js

import * as rentalsJson from "@api/rentals.json";

const COMMUNITY_CATEGORIES = ["Condo", "Townhouse", "Apartment"];

export const load = () => {
  const rentals = rentalsJson.data.map(value => {
    const { id, attributes } = value;
    let type = "Community";

    if (!COMMUNITY_CATEGORIES.includes(attributes.category)) {
      type = "Standalone";
    }

    return { id, type, ...attributes };
  });

  if (rentals) {
    return { rentals };
  }
};
```

---

> The `load` function is a reserved word and is used to load the relevant data for

Reviewer: ```suggestion
The `load` function is used to load the relevant data for
```

"reserved word" is a bit confusing – reserved keywords in languages are words that have meaning and you cannot use to name e.g. your variables, for example `function` or `async` in JS. This is just Svelte's API.

    if (key.includes(params.slug)) {
      rental = rentals[key];
    }
  }
  if (rental) {
    const { data } = await rental();
    const { attributes } = data;
    let type;

    if (COMMUNITY_CATEGORIES.includes(attributes.category)) {
      type = "Community";
    } else {
      type = "Standalone";
    }

    return { rental: { ...attributes, type } };
  }
}
```

We are using the `load` function just as we did in the last section, but this
time `load` receives an object with the attribute `{ params }` and inside that
object we can find all of the parameters from the route, in this case we are
just interested in the `slug` as this is what we will use to find the current
rental.

Because of the way that the imports are handled on the server, we aren’t able to
access a file using a dynamic name directly, so what we do here instead is get
all of the files from `@api/rentals/*.json` and then go through that array to
find the rental of the route we are currently on.

---

This isn’t ideal, and the loading would normally be done by our back-end, but
for the limited number of rentals we have, this works nicely.

---

We can then create a page to display this data

> ```js

Reviewer: ```suggestion
```html
```

      <span>Type:</span>
      {rental.type} – {rental.category}
    </div>
    <div class="detail location">
      <span>Location:</span>
      {rental.city}
    </div>
    <div class="detail bedrooms">
      <span>Number of bedrooms:</span>
      {rental.bedrooms}
    </div>
    <div class="detail description">
      <p>{rental.description}</p>
    </div>
  </div>

  <Map
    lat={rental.location.lat}
    lng={rental.location.lng}
    zoom="12"
    width="894"
    height="600"
    alt="A map of {rental.title}"
    styleClass="large"
  />
</article>
```

As before, we’ll use a `script` tag with `context=”module”` to import things
that won’t change between instances of the page, and a `script` without the
context for the attributes that can change. You’ll notice here that we are
expanding the `data` object and assigning `data.rental` to it’s own property to
make it easier to reference in the template. We will implement the “Share on
Twitter” button later so this is just a placeholder for now.

We also need to be able to add a class to the `<Map>` component so that it can
be styled from the parent, so to do this we will add a new attribute to the
component.

> ```diff-js

Reviewer: ```suggestion
```diff-html
```

Author: This has a similar issue to the js, the diff styling can be added, but the HTML styling is missing

// components/map.svelte

  ...
  export let alt = `Map image at coordinates ${lat},${lng}`;
+ export let styleClass = '';

  $: src = `${MAPBOX_API}/${lng},${lat},${zoom}/${width}x${height}@2x?access_token=${token}`;
</script>

- <div class="map">
+ <div class="map {styleClass}">
    <img {alt} {src} {width} {height} />
  </div>
...
```

This allows us to add a class to the component from the parent, but because the
class encapsulation doesn’t work here, we still need to use a global class (like
the ones we defined in the `global.css`) or use the `:global()` css selector
which means we need to be careful to make this selector specific so it doesn’t
leak into other styles. This emphasised how useful Glimmer’s `...attributes`
syntax is, where you can simply pass any attributes you want from parent to
child component.

To give users access to the individual rental detail page, we will update the
`<Rental>` component to contain a link to the rental detail page

> ```diff-js

Reviewer: ```suggestion
```diff-html
```

Author: This has a similar issue to the js, the diff styling can be added, but the HTML styling is missing

  // Assert the rental detail area is visible
  await expect(page.locator('.rental.detailed')).toBeVisible();
});
```

### Service injection

[Repo](https://github.com/mainmatter/sveltekit-super-rentals/tree/2.2-service-injection)

Let’s continue by replacing that placeholder “Share to Twitter” button we
created in the last section with a real, working button that can share a tweet
to Twitter. First, we’ll create a new component to house all of this Twitter
logic

```js
// components/share-button.svelte

<script>
  import { onMount } from 'svelte';

  const TWEET_INTENT = 'https://twitter.com/intent/tweet';

  export let text
  export let hashtags
  export let via

  let shareURL;

  onMount(() => {
    let url = new URL(TWEET_INTENT);

    url.searchParams.set('url', window.location.href);

    if (text) {
      url.searchParams.set('text', text);
    }

    if (hashtags) {
      url.searchParams.set('hashtags', hashtags);
    }

    if (via) {
      url.searchParams.set('via', via);
    }

    shareURL = url.href;
  });
</script>

<a href={shareURL} target="_blank" rel="external nofollow noopener noreferrer" class="share button">
  <slot />
</a>
```

---

> The `onMount` function here is very similar to Glimmer’s `constructor`, this is

Reviewer: I think this makes it sound like the constructor is somehow a concept of Glimmer. I'd consider changing to sth. like

```suggestion
The `onMount` function here is very similar to Glimmer components' constructors, this is
```

where you store the logic that needs to run before the component is rendered, in
this scenario we will be populating the URL that we will need to create a tweet
to share the rental we are looking at.

---

We will replace the placeholder with this new component on the `rentals/[slug]`
page

> ```diff-js

Reviewer: ```suggestion
```diff-html
```

+  const href = await page.getAttribute('.share.button', 'href');
+  const tweetURL = new URL(href);
+  expect(tweetURL.host).toBe('twitter.com');
+  expect(tweetURL.searchParams.get('url')).toBe(`${[page.url()]}`);

  // Assert it has taken you to the rental page
  await expect(page).toHaveURL('/rentals/grand-old-mansion');
  ...
```

In the Ember-super-rentals tutorial, there are steps about how to get the Router
service to allow these tests to pass, but the Router service doesn’t exist in
SvelteKit in the same way it does in Ember, so we don’t require anything special
for the tests; Playwright does give us access to the `page` attribute in the
tests, so it’s nice and simple to grab the current URL from there to compare
instead of needing to go through the router.

### Ember Data

[Repo](https://github.com/mainmatter/sveltekit-super-rentals/tree/2.3-ember-data)

Because we are not using Ember Data, this part of the tutorial can be skipped

### Provider components

[Repo](https://github.com/mainmatter/sveltekit-super-rentals/tree/2.4-provider-components)

And to round things off, we will be adding the ability to filter our rentals

> based on a text search

Reviewer 2: ```suggestion
based on a text search.
```

Because we essentially already have the index route as a component, we won’t
need to separate out the rentals into their own component as we can easily keep
track of the searchQuery directly in the index route.

```js
// components/rentals-filter.svelte

<script>

> 

Reviewer: ```suggestion
```

  export let rentals;
  export let query;

  // we want to make the query lowercase and remove trailing/leading spaces
  $: sanitisedQuery = query.toLowerCase().trim();

  $: results = query
    ? rentals.filter(rental => {
        return rental.title.toLowerCase().includes(sanitisedQuery);
      })
    : rentals;
</script>

<slot {results} />
```

So here we are expecting an array of rentals and a query, then doing some very
basic sanitisation on the query, then if there is a query (it’s not blank or
undefined) we will filter the rentals based on the sanitisedQuery and then pass
back the results. This is very similar to Ember in the way we can pass a
property back to the parent component by setting `{{yield results}}`.

We can then update the index page to keep track of the query and utilise the new
`<RentalsFilter>` component to display only the filtered rentals

> ```diff-js

Reviewer: ```suggestion
```diff-html
```

Author: This has a similar issue to the js, the diff styling can be added, but the HTML styling is missing

  expect(page.locator('.rentals')).toBeVisible();
  expect(page.locator('.rentals input')).toBeVisible();
  expect(page.locator('.rentals .results')).toBeVisible();
  expect(page.locator('.rentals .results li')).toHaveCount(3);

  // Fill input
  await page.locator('input').fill('Mansion');

  // assert that there is only one result and that it's title includes the word "mansion"
  expect(page.locator('.rentals .results li')).toHaveCount(1);
  expect(await page.textContent('.rentals .results li h3')).toBe('Grand Old Mansion');

  // Fill input with new search term
  await page.locator('input').fill('DownTown');

  // assert that there is only one result and that it's title includes the word "downtown"
  expect(page.locator('.rentals .results li')).toHaveCount(1);
  expect(await page.textContent('.rentals .results li h3')).toBe('Downtown Charm');
});
```

And there we have it, SvelteKit Super Rentals! (It's not covered as part of this
tutorial, but you can see the final version deployed to Netlify
[here](https://sveltekit-super-rentals.netlify.app/))

### Final Thoughts

I really enjoyed getting stuck in with SvelteKit and I can certainly see the

> appeal of it, it’s a very lightweight framework and in a lot of ways it stays

Reviewer 2: ```suggestion
appeal of it. It’s a very lightweight framework and in a lot of ways it stays
```

> very close to native javascript, which makes me think it would be a really great

Reviewer 2: ```suggestion
very close to native JavaScript, which makes me think it would be a really great
```

first framework for people that are new to front-end development, because it
gives you a great understanding of how javascript works in a lot of scenarios,

> which will only help when trying out other frameworks in the future. Which I

Reviewer 2: ```suggestion
which will only help when trying out other frameworks in the future. I feel this
```

> feel is almost the opposite of Ember, there are a lot of areas of the framework

Reviewer 2: ```suggestion
is almost the opposite of Ember: there are a lot of areas of the framework
```

where Ember does the heavy lifting for you, which is great for large
applications, but possibly not the best place to start if you want to build a
base knowledge of the underlying technologies.

There were a few areas I did think SvelteKit could learn from Ember though. I
find writing tests is much more enjoyable in Ember, and I think that shines a

> light on SvelteKit’s test suite as a whole, Ember’s qunit integration allows for

Reviewer 2: ```suggestion
light on SvelteKit’s test suite as a whole. Ember’s qunit integration allows for
```
